src/server/game/Texts/ChatTextBuilder.cpp

```diff
@@ -27,15 +27,57 @@
 namespace Trinity
 {
 ChatPacketSender::ChatPacketSender(ChatMsg chatType, ::Language language, WorldObject const* sender, WorldObject const* receiver,
-    std::string message, uint32 achievementId /*= 0*/, LocaleConstant locale /*= LOCALE_enUS*/)
-    : Type(chatType), Language(language), Sender(sender), Receiver(receiver), Text(std::move(message)), AchievementId(achievementId), Locale(locale)
+    std::string message, uint32 achievementId /*= 0*/, LocaleConstant locale /*= LOCALE_enUS*/, uint32 broadcastTextId /*= 0*/, uint16 emoteId /*= 0*/, uint32 soundKitId /*= 0*/, SoundKitPlayType soundKitPlayType /*= SoundKitPlayType::Normal*/, uint32 playerConditionId /*= 0*/)
+    : Type(chatType), Language(language), Sender(sender), Receiver(receiver), Text(std::move(message)), AchievementId(achievementId), Locale(locale), PlayerConditionID(playerConditionId)
 {
     UntranslatedPacket.Initialize(Type, Language, Sender, Receiver, Text, AchievementId, "", Locale);
     UntranslatedPacket.Write();
+
+    if (sender && sender->IsUnit() && emoteId)
+    {
+        EmotePacket.emplace();
+        EmotePacket->Guid = sender->GetGUID();
+        EmotePacket->EmoteID = emoteId;
+        EmotePacket->Write();
+    }
+
+    if (soundKitId)
+    {
+        if (soundKitPlayType == SoundKitPlayType::Normal)
+        {
+            SoundPacket.emplace();
+            SoundPacket->SoundKitID = soundKitId;
+            SoundPacket->BroadcastTextID = broadcastTextId;
+            SoundPacket->SourceObjectGuid = sender ? sender->GetGUID() : ObjectGuid::Empty;
+            SoundPacket->Write();
+        }
+        else if (soundKitPlayType == SoundKitPlayType::ObjectSound)
+        {
+            ObjectSoundPacket.emplace();
+            ObjectSoundPacket->TargetObjectGUID = receiver ? receiver->GetGUID() : ObjectGuid::Empty;
+            ObjectSoundPacket->SourceObjectGUID = sender ? sender->GetGUID() : ObjectGuid::Empty;
+            ObjectSoundPacket->SoundKitID = soundKitId;
+            if (receiver)
+                ObjectSoundPacket->Position = receiver->GetWorldLocation();
+            ObjectSoundPacket->BroadcastTextID = broadcastTextId;
+            ObjectSoundPacket->Write();
+        }
+    }
 }
 
 void ChatPacketSender::operator()(Player const* player) const
 {
+    if (!player->MeetPlayerCondition(PlayerConditionID))
+        return;
+
+    if (SoundPacket)
+        player->SendDirectMessage(SoundPacket->GetRawPacket());
+    else if (ObjectSoundPacket)
+        player->SendDirectMessage(ObjectSoundPacket->GetRawPacket());
```

Comment:
We could merge `SoundPacket` and `ObjectSoundPacket` and use `ServerPacket` instead. Due to inheritance magic `GetRawPacket()` is valid for both and polymorphism does the rest